File: 2020/day05/day05.go
// --- Day 5: Binary Boarding ---
// https://adventofcode.com/2020/day/5

package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strings"
)

type BoardingPass struct {
	Data string
}

func (bp *BoardingPass) SeatID() int {
	return (bp.findRow() * 8) + bp.findColumn()
}

func (bp *BoardingPass) findRow() int {
	return bp.bsearch(bp.Data[0:7], "F", "B")
}

func (bp *BoardingPass) findColumn() int {
	return bp.bsearch(bp.Data[7:10], "L", "R")
}

func (bp *BoardingPass) bsearch(value, lowerKey, upperKey string) int {
	lower := 0
	upper := int(math.Exp2(float64(len(value))) - 1)
	split := strings.Split(value, "")

	for _, key := range split {
		if key == lowerKey {
			upper = int(math.Floor(float64(upper-lower) / 2.0))
			upper += lower
		} else if key == upperKey {
			lower += int(math.Ceil(float64(upper-lower) / 2.0))
		}
	}

	return lower
}

func main() {
	boardingPasses, err := readBoardingPasses()
	if err != nil {
		panic(err)
	}

	sort.SliceStable(boardingPasses, func(i, j int) bool {
		return boardingPasses[i].SeatID() < boardingPasses[j].SeatID()
	})

	part1 := boardingPasses[len(boardingPasses)-1].SeatID()

	var part2 int
	for index, bp := range boardingPasses {
		if boardingPasses[index+1].SeatID() == bp.SeatID()+2 {
			part2 = bp.SeatID() + 1
			break
		}
	}

	fmt.Println("--- Day 5: Binary Boarding ---")
	fmt.Println("Part 1:", part1)
	fmt.Println("Part 2:", part2)
}

func readBoardingPasses() ([]BoardingPass, error) {
	boardingPasses := make([]BoardingPass, 0)
	file, err := os.Open("./day05.txt")
	if err != nil {
		return boardingPasses, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		boardingPasses = append(boardingPasses, BoardingPass{Data: scanner.Text()})
	}

	if err := scanner.Err(); err != nil {
		return boardingPasses, err
	}

	return boardingPasses, nil
}

File: 2020/day05/day05.rb
# --- Day 5: Binary Boarding ---
# https://adventofcode.com/2020/day/5

class BinaryBoarding
  class BoardingPass
    attr_reader :data

    def initialize(data)
      @data = data.strip
    end

    def to_s
      "#{data} - #{find_row} * 8 + #{find_column} = #{seat_id}"
    end

    def seat_id
      (find_row * 8) + find_column
    end

    def bsearch(value, lower:, upper:)
      boundary = 0..((2**value.length) - 1)
      value.split("").each do |key|
        if key == lower
          lower_bound = boundary.min
          upper_bound = (boundary.max - lower_bound) / 2
          upper_bound += lower_bound
          boundary = lower_bound..upper_bound
        elsif key == upper
          upper_bound = boundary.max
          lower_bound = boundary.min
          lower_bound += ((upper_bound - lower_bound) / 2.0).ceil
          boundary = lower_bound..upper_bound
        end
      end
      boundary.min
    end

    def find_row
      bsearch(data[0..6], lower: "F", upper: "B")
    end

    def find_column
      bsearch(data[7..-1], lower: "L", upper: "R")
    end
  end

  attr_reader :boarding_passes

  def initialize
    @boarding_passes = ::File.foreach("./day05.txt").map do |data|
      BoardingPass.new(data)
    end
  end

  def part1
    @boarding_passes.max { |bp1, bp2| bp1.seat_id <=> bp2.seat_id }
  end

  def part2
    ordered = @boarding_passes.sort_by { |bp| bp.seat_id }
    prev_seat = ordered.detect.with_index do |bp, index|
      ordered[index+1].seat_id == bp.seat_id + 2
    end

    prev_seat.seat_id + 1
  end
end

bb = BinaryBoarding.new
puts "--- Day 5: Binary Boarding ---"
puts "Part 1: #{bb.part1}"
puts "Part 2: #{bb.part2}"
